src/com/stuypulse/stuylib/file/FRCLogger.java

```diff
@@ -1,71 +1,164 @@
+     * The Loggable interface should be implemented by a class that can 
+     * return a useful piece of data that can be called on demand to be written
+     * into the log.
      */
     public interface Loggable {
-        public String getLog();
+        /**
+         * Implement this method with something that reports
+         * if any data from the class should be fed to the logger
+         * this iteration.
+         * 
+         * @return if any data from the class will be fed to the logger.
+         */
+        public boolean logThisIteration();
+
+        /**
+         * Implement this method with something that
+         * reports the level of severity the data should be
+         * logged at.
+         * 
+         * @return the level of the data that will be fed.
+         */
+        public default Level getLogLevel() {
+            return Level.INFO;
+        }
+
+        /**
+         * Implement this method with something that 
+         * reports the message or data that should be fed
+         * into the logger as a string.
+         * 
+         * @return the data to be fed into the logger.
+         */
+        public String getLogData();
+        
     }
 
     private Logger mLogger;
     private FileHandler mFileHandler;
 
+    private ArrayList<Loggable> loggables;
+
     /**
-     * Open FRCLogger
+     * Open a new FRCLogger
      * 
-     * @param file file name
+     * @param path Name of the director to store the log file to.
+     * @param name Name of the file to write logs to.
      */
-    FRCLogger(String file) {
+    FRCLogger(String path, String name) {
         mLogger = Logger.getLogger(FRCLogger.class.getName());
 
         try {
-            mFileHandler = new FileHandler("./Logs/" + file + ".log");
-        } catch (Exception e) {
+            mFileHandler = new FileHandler(path + "/Logs/" + name + ".log");
+        } catch (IOException e) {
+            logError(this, e);
             e.printStackTrace();
         }
 
         mFileHandler.setFormatter(new SimpleFormatter());
         mLogger.addHandler(mFileHandler);
+
+        loggables = new ArrayList<Loggable>();
     }
 
     /**
-     * Log Loggable Class
+     * Gets the list of registered Loggables.
      * 
-     * @param in Loggable Class
+     * @return all registered Loggables as an array.
      */
-    public void log(Loggable in) {
-        mLogger.info(in.getClass().getName().toUpperCase() + ":\n" + in.getLog());
+    public Loggable[] getLoggables() {
+        return (Loggable[])loggables.toArray();
     }
 
     /**
-     * Log String as Info
+     * Registers a new Loggable that upon call will put the next
+     * getLogData() at level getLogLevel() to the log if it will
+     * logThisIteration().
      * 
-     * @param info the information
+     * @param loggable Class implementing Loggable
      */
-    public void logInfo(String info) {
-        mLogger.info(info);
+    public void registerLoggable(Loggable loggable) {
```

Comment:
maybe rename to addLoggable
register seems a bit odd

Reply:
Register is to reminisce of events and event handlers for people who come from javascript.  Also it semi makes sense in that context.  Also we want our code to be UNIQUE, not like all those generic adds.

Reply:
-Ivan from Ayan's account